// chrome-app-developer-tool-client/chromeharnessclient.js
/**
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
 */
var fs = require('fs');
var HarnessClient = require('./harnessclient');
var ChromePushSession = require('./chromepushsession');

function ChromeHarnessClient(target) {
    HarnessClient.call(this, target);
}
ChromeHarnessClient.detectAdbTargets = HarnessClient.detectAdbTargets;
ChromeHarnessClient.prototype = Object.create(HarnessClient.prototype);
ChromeHarnessClient.prototype.constructor = ChromeHarnessClient;

ChromeHarnessClient.prototype.createPushSession = function(dir) {
    return new ChromePushSession(this, dir);
};

ChromeHarnessClient.prototype.buildApk = function(appId, appType, signingOpts, outputPath, /* optional */ manifestEtag) {
    var query = {};
    if (manifestEtag) {
        query['manifestEtag'] = manifestEtag;
    }
    var payload = {
        'keyPassword': signingOpts.keyPassword
    };
    if (signingOpts.privateKeyPath) {
        payload['publicKeyData'] = fs.readFileSync(signingOpts.certificatePath, 'base64');
        payload['privateKeyData'] = fs.readFileSync(signingOpts.privateKeyPath, 'base64');
    } else {
        payload['storeData'] = fs.readFileSync(signingOpts.keyStorePath, 'base64');
        payload['storeType'] = signingOpts.storeType;
        payload['storePassword'] = signingOpts.storePassword;
        payload['keyAlias'] = signingOpts.keyAlias;
    }
    var opts = {
        appId: appId,
        appType: appType,
        json: payload,
        query: query,
        saveResponseToFile: outputPath
    };
    return this.doRequest('POST', 'buildapk', opts);
};

module.exports = ChromeHarnessClient;
